### PR TITLE
Updating the conflict.py

### PR DESCRIPTION
**Pull request by stannida**
Updating the ConflictManager (conflict check and conflict time aware)

### CODE REVIEW

File: metrics/conflict.py [hunk 1]
```diff
@@ -129,30 +126,34 @@ def __calculate_time_diffs(self, elegible_actions):
         # first calculate the times for all cases. This will produce some errors because
         # the shift is not aware of the tokens (revision times should belong to the same
         # token). This errors are removed in the next lines
-        df['time_diff'] = df['rev_time'] - df.shift(2)['rev_time']
-
+        
+        # changed:  instead of shifting by 2, shifting by 1
+        df['time_diff'] = 0
```
Comment (robertour) [inline metrics/conflict.py]: this line is unnecessary, line 132 and 145 will override it anyway

File: metrics/conflict.py [hunk 2]
```diff
@@ -31,11 +32,12 @@ def calculate(self):
...
 
-        print('Calculate time differences of undos')
-        elegible = self.__calculate_time_diffs(elegible)
-
         print('Get the conflicts')
         self.__conflicts = self.__get_conflicts(elegible)
+        
+        print('Calculate time differences of undos')
+        elegible = self.__calculate_time_diffs(elegible)
+        elegible.head()
```
Comment (robertour) [inline metrics/conflict.py]: remove this line

File: metrics/conflict.py [hunk 3]
```diff
@@ -163,10 +164,14 @@ def __get_conflicts(self, df):
...
-                (df['editor'] != df.shift(1)['editor']) &
-                (df['editor'] == df.shift(2)['editor']))
+        # what it should be:
+        # the token is the same as the previous
+        # out editor is different from in or vice versa
+        # changed: we do not consider a conflict only those actions, where the revision is made 
+        #by the same user or the first insertion.
+        df['is_a_conflict'] = ((df['token_id'] == df.shift(1)['token_id']) &
+                (df['editor'] != df.shift(1)['editor']))
+        return df
```
Comment (robertour) [inline metrics/conflict.py]: the logic seems right, try to return the boolean series directly 
```
return ((df['token_id'] == df.shift(1)['token_id']) &
                (df['editor'] != df.shift(1)['editor']))
```

File: metrics/conflict.py [hunk 4]
```diff
@@ -186,26 +191,28 @@ def calculate_token_conflict_score(self, df, conflicts, base=3600):
         2. Calculating the log(t, base=3600) soften the curve so that the values are not so extreme. 
         Moreover, it sets 1 hour (3600 secs) as the decisive point in which an undo is more relevant.
         """
+        #changed: time_diff is not calculated for the first insertion so we can emit this checking
         df['conflict'] = np.nan
-        df.loc[conflicts, 'conflict'] = np.log(
-            base) / np.log(df.loc[conflicts, 'time_diff'].astype('timedelta64[s]') + 2)
+        df.loc[df['is_a_conflict']==True, ['conflict']] = np.log(
```
Comment (robertour) [inline metrics/conflict.py]: if the return value of  `def __conflicts` is a series, then you could `conflicts` instead of  `is_a_conflict`.

File: metrics/conflict.py (hunk 1 above)
Comment (stannida) [reply metrics/conflict.py]: done

File: metrics/conflict.py [hunk 5]
```diff
@@ -186,26 +191,28 @@ def calculate_token_conflict_score(self, df, conflicts, base=3600):
...
-        and divide them by the summ of all elegible actions (i.e. actions that have the potential
+        and divide them by the sum of all elegible actions (i.e. actions that have the potential
         of being undos)
         """
 
         if (self.elegible_actions.shape[0] == 0):
             return 0
         else:
-            return (self.elegible.loc[self.__conflicts, 'conflict'].sum() /
+            return (self.elegible.loc[self.__conflicts['is_a_conflict'], 'conflict'].sum() /
```
Comment (robertour) [inline metrics/conflict.py]: same should be here

File: metrics/conflict.py [hunk 6]
```diff
@@ -227,16 +234,8 @@ def get_conflict_score_per_editor(self):
         joined = confs_n.join(confs_ed).join(actions)
 
         # calculate the score of the editor dividing conflicts / actions
-        joined['conflict_score'] = joined['conflict'] / joined['action']
+        joined['conflict_score_ed'] = joined['conflict'] / joined['action']
```
Comment (robertour) [inline metrics/conflict.py]: why the change of name? doesn't this affect the jupyter notebooks?

File: metrics/conflict.py [hunk 7]
```diff
@@ -227,16 +234,8 @@ def get_conflict_score_per_editor(self):
...
-        return joined.sort_values('conflict_score', ascending=False)
-
-    # c_t = np.log(3600) / (
-    #     np.log(
-    #         dups_dated.loc[conflicts,['token_id','time_diff']].groupby(
-    #             'token_id').sum().astype('timedelta64[s]') + 2
-    #     ))
-
-    # c_t.sum()
+        return joined.sort_values('conflict', ascending=False)
```
Comment (robertour) [inline metrics/conflict.py]: here you are sorting by `conflict` and before was sorted by the `conflict_score`. I am not sure which does make more sense, or if they are even differrent, ?

**Review by robertour — CHANGES_REQUESTED** [review]
The logic of the changes is correct, there are just a few minor issues to address.

DISCLAIMER: I have never used this reviewing feature in github
